core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1671,7 +1671,6 @@ class Log(@volatile var dir: File,
   private[log] def delete() {
     maybeHandleIOException(s"Error while deleting log for $topicPartition in dir ${dir.getParent}") {
       lock synchronized {
-        checkIfMemoryMappedBufferClosed()
         removeLogMetrics()
```

Comment:
I am wondering if it would be more intuitive not to set `isMemoryMappedBufferClosed = true` in `delete()`. 

The reason is that, if we don't call `checkIfMemoryMappedBufferClosed()` in delete(), it suggests that delete() can be done anytime after close() is called, which is actually the case in `logManager.replaceCurrentWithFutureLog()` which calls `addLogToBeDeleted(sourceLog)` after `sourceLog.close()`. Then the intuition is that, if delete() can be done without considering the state (i.e. isMemoryMappedBufferClosed) of log, it probably should not change the state of log as well.

Another intuition is that, it seems weird if delete() sets `isMemoryMappedBufferClosed` to true and yet delete() can be called again without triggering `KafkaStorageException`.

Reply:
It's a good question. It's a little unclear what's the best path as we don't generally require `close` to be called before `delete` for the various log/index classes. As a result, we have one case where we don't call `Log.close` at all (the `LogManager.asyncDelete` path) and rely on the `Log.delete` method to release resources. If we don't set `isMemoryMappedBufferClosed` in the `delete` method, then it would stay `false` for this path.

As it happens, the fact that we don't close the file handles before we rename the directory in `LogManager.asyncDelete` is one reason why things break on Windows.

Reply:
Regarding the design for the `Log.java` API, it seems intuitive to only allow `delete()` after `close()` or `closeHandlers()` is invoked. `delete()` will only delete files on disk without reading or writing anything. And it seems reasonable to close a log before deleting it in e.g. `LogManager.asyncDelete()`.

Reply:
Seems like this needs a bit more discussion. I propose we revert the change in 2.1 to close the index during Log.close so that the behaviour remains the same as in 2.0. In trunk, we can make the changes after we agree to them. How does that sound?

Reply:
Sure. Sounds good. Thanks for the patch! I will take care of the revert in the 2.1 branch.